stage #2.5/module #3 Spring Core/#5.1. Annotation-based configuration/content.md

```diff
@@ -0,0 +1,470 @@
+
+    private MusicMagazine hardRockMusicMagazine;
+
+    @Autowired
+    public void setHardRockMusicMagazine(@MusicGenre("HardRock") MusicMagazine hardRockMusicMagazine) {
+        this.hardRockMusicMagazine = hardRockMusicMagazine;
+    }
+
+    // ...
+}
+```
+
+Next, you can provide the information for the candidate bean definitions. You can add `<qualifier/>` tags as
+sub-elements of the `<bean/>` tag and then specify the `type` and `value` to match your custom qualifier annotations.
+The type is matched against the fully-qualified class name of the annotation. Alternately, as a convenience if no risk
+of conflicting names exists, you can use the short class name. The following example demonstrates both approaches:
+
+```xml
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns="http://www.springframework.org/schema/beans"
+       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xmlns:context="http://www.springframework.org/schema/context"
+       xsi:schemaLocation="http://www.springframework.org/schema/beans
+        https://www.springframework.org/schema/beans/spring-beans.xsd
+        http://www.springframework.org/schema/context
+        https://www.springframework.org/schema/context/spring-context.xsd">
+
+    <context:annotation-config/>
+
+    <bean class="com.somepackage.MusicMagazineImpl">
+        <qualifier type="MusicGenre" value="HeavyMetal"/>
+        <!-- inject any dependencies required by this bean -->
+    </bean>
+
+    <bean class="com.somepackage.MusicMagazineImpl">
+        <qualifier type="com.somepackage.MusicGenre" value="HardRock"/>
+        <!-- inject any dependencies required by this bean -->
+    </bean>
+
+    <bean id="musicRatingService" class="com.somepackage.MusicRatingService"/>
+
+</beans>
+```
+
+## @Resource
+
+Spring also supports injection by using the JSR-250 `@Resource` annotation (`javax.annotation.Resource`) on fields or
+bean property setter methods.
+
+`@Resource` takes a name attribute. By default, Spring interprets that value as the bean name to be injected. In other
+words, it follows by-name semantics, as demonstrated in the following example:
+
+```java
+public class MusicRatingService {
+
+    private MusicRatesFinder musicRatesFinder;
+
+    @Resource(name = "myMusicRatesFinder")
+    public void setMusicRatesFinder(MusicRatesFinder musicRatesFinder) {
+        this.musicRatesFinder = musicRatesFinder;
+    }
+}
+```
+
+If no name is explicitly specified, the default name is derived from the field name or setter method. In case of a
+field, it takes the field name. In case of a setter method, it takes the bean property name.
+
+In the exclusive case of `@Resource` usage with no explicit name specified, and similar to `@Autowired`, `@Resource`
+finds a primary type match instead of a specific named bean and resolves well known resolvable dependencies: the
+`BeanFactory`, `ApplicationContext`, `ResourceLoader`, `ApplicationEventPublisher`, and `MessageSource` interfaces.
```

Comment:
I didn't get what the difference between `@Qualifier` and `@Resource`.
Add smth like:
`- @Autowired is used together with @Qualifier , it is the same as the @Resource
-@Resource only supports for fields and setter injection while @Autowired supports fields , setter ,constructors and multi-argument methods injection.
-It is suggested to use @Resource for fields and setter injection. Stick with @Qualifier and @Autowired for constructor or a multi-argument method injection.`

Reply:
There was no intention to provide a difference between @Qualifier and @Resource, but I can add your points, thanks a lot for them.

Reply:
Reading all types of annotation I start to analize what the difference of describing annotation and previous ones.
So, I believe prividing difference or sprecial things about annotation may help to understand and memorize them better. 
Thank you for adding this information